arlas-core/src/main/java/io/arlas/server/services/ExploreServices.java

```diff
@@ -350,25 +350,38 @@ protected void applyProjection(Projection projection, FluidSearch fluidSearch, O
+    public AggregationResponse formatAggregationResult(MultiBucketsAggregation aggregation, AggregationResponse aggregationResponse, CollectionReference collection, List<Aggregation> aggregationsRequest, int aggTreeDepth) {
         aggregationResponse.name = aggregation.getName();
         if (aggregationResponse.name.equals(FluidSearch.TERM_AGG)) {
             aggregationResponse.sumotherdoccounts = ((Terms) aggregation).getSumOfOtherDocCounts();
         }
+        RawGeometries rawGeometries = aggregationsRequest.size() > aggTreeDepth ? aggregationsRequest.get(aggTreeDepth).rawGeometries : null;
+        List<AggregatedGeometryEnum> aggregatedGeometries = aggregationsRequest.size() > aggTreeDepth ? aggregationsRequest.get(aggTreeDepth).aggregatedGeometries : null;
         aggregationResponse.elements = new ArrayList<AggregationResponse>();
         List<MultiBucketsAggregation.Bucket> buckets = (List<MultiBucketsAggregation.Bucket>) aggregation.getBuckets();
         buckets.forEach(bucket -> {
             AggregationResponse element = new AggregationResponse();
             element.keyAsString = bucket.getKeyAsString();
-            if (aggregationResponse.name.startsWith(FluidSearch.GEOHASH_AGG)) {
+            /** if it is a `geohash` aggregation type, we set the GEOHASHCENTER and GEOHASH aggregated geometries if they're requested **/
+            if (aggregationResponse.name.equals(FluidSearch.GEOHASH_AGG)) {
                 GeoPoint geoPoint = getGeohashCentre(element.keyAsString.toString());
                 element.key = geoPoint;
-                if (aggregationResponse.name.equals(FluidSearch.GEOHASH_AGG)) {
-                    // return the centroid of the geohash
-                    element.geometry = new Point(geoPoint.getLon(), geoPoint.getLat());
-                } else {
-                    // return the Extent of the geohash
-                    element.geometry = createPolygonFromRectangle(GeohashUtils.decodeBoundary(element.keyAsString.toString(), SpatialContext.GEO));
+                if (!CollectionUtils.isEmpty(aggregatedGeometries)) {
+                    aggregatedGeometries.stream().filter(g -> g == AggregatedGeometryEnum.GEOHASHCENTER || g == AggregatedGeometryEnum.GEOHASH).forEach(g -> {
+                        ReturnedGeometry returnedGeometry = new ReturnedGeometry();
+                        returnedGeometry.reference = g.value();
+                        returnedGeometry.isRaw = false;
+                        if (g == AggregatedGeometryEnum.GEOHASH) {
+                            returnedGeometry.geometry = createPolygonFromRectangle(GeohashUtils.decodeBoundary(element.keyAsString.toString(), SpatialContext.GEO));
+                        } else {
+                            returnedGeometry.geometry = new Point(geoPoint.getLon(), geoPoint.getLat());
+                        }
+                        if (element.geometries == null) {
```

Comment:
It looks like it is repeated, can that be done at the beginning?

Reply:
It's repeated it because I wanted to create the list only when it can be populated with geometries; Otherwise I keep it null so that I don't have the attribute in the response